### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by mzgubic**
Some questions and points for discussion are highlighted in my initial code review

### CODE REVIEW

File: src/CachedCalls.jl
Comment (mzgubic) [inline src/CachedCalls.jl]: This needs to change to allow the user to specify their own directory.

One idea is to use environment variables which can be specified in the script that the user runs. Is that a good way to go about it? On downside is that the ENV is stored as a string which makes directory manipulation somewhat more fiddly.

File: src/CachedCalls.jl
Comment (mzgubic) [inline src/CachedCalls.jl]: This function is very fiddly because the kwargs can be specified in three different ways (see tests) and that results in different expressions that all need to be parsed to the same thing. Particularly annoying is that the `f(;kwarg)` syntactic sugar adds the need for the `keep` kwarg.

I would love to this in a different way

File: README.md
Comment (mattBrzezinski) [inline README.md]: Newline before/after?

```suggestion

```julia
using CachedCalls

@cached_call f(args; kwargs)
```

```

File: src/CachedCalls.jl
Comment (mattBrzezinski) [reply src/CachedCalls.jl]: You could do something like this, where you set a default and give the user an option to override?
Also combine that with maybe an env var for it to persist?

> https://github.com/JuliaCloud/AWS.jl/blob/3dee0519b568c49232ae0b753ca3d5a5fb01d484/src/AWS.jl#L31-L67

File: src/CachedCalls.jl
Comment (mattBrzezinski) [inline src/CachedCalls.jl:115]:
```suggestion

function _extract_kwargs(a::AbstractArray; keep_args=false)
```

File: src/CachedCalls.jl
Comment (mattBrzezinski) [inline src/CachedCalls.jl:101]:
```suggestion

function _extract_kwargs(ex::Expr; keep_args=false)
```

File: src/CachedCalls.jl
Comment (mattBrzezinski) [inline src/CachedCalls.jl]: I think there's a better way to create a string here, I'm not familiar w/ quote though.

File: src/CachedCalls.jl
Comment (nickrobinson251) [reply src/CachedCalls.jl]: maybe put it under `.julia/` like DataDeps does? 
https://github.com/oxinabox/DataDeps.jl/blob/98e2f708a7fb798fce67269d2b7a791441370fd9/src/locations.jl#L4-L19

e.g. default could be `joinpath(first(Base.DEPOT_PATH), "cachedcalls")`

Comment (mzgubic) [reply src/CachedCalls.jl]: Thanks both, this was exactly what I was looking for!

**Comment by mzgubic** [comment]
Are we ok with this?

**Comment by mzgubic** [comment]
ping

**Comment by nickrobinson251** [comment]
> Are we ok with this?

Are you asking me? I don't have nearly enough context of the use-case to answer that

Also is this for Invenia's use? If so should it be in the Invenia org? And can someone you're working with review?

**Comment by mzgubic** [comment]
I am asking into the void 😂 

This project isn't directly related to a squad, thought I think we could use it in several squads to speed up development. It's just a fun thing to work on as a low priority. I am trying to follow contributors guide and not merge before approval, that's all. I should have explained that more clearly, my bad!

**Review by mattBrzezinski — APPROVED** [review]

**Comment by mattBrzezinski** [comment]
LGTM! Merge it, tag it, and post in the Julia Slack about it! :)

**Pull request merged**
